unified_addrs.py

```diff
@@ -0,0 +1,166 @@
+            assert [1, 2, 3].count(b) > 0, "receiver type " + str(b) + " not recognized"
+            s = 1
+        elif s == 1:
+          receiver_len = b
+          expected_len == {1: 20, 2: 43, 3: 43}.get(receiver_type)
+            if expected_len is not None:
+                assert receiver_len == expected_len, "incorrect receiver length"
+        elif s == 2:
+            if len(acc) < receiver_len:
+                acc.append(b)
+            
+            if len(acc) == receiver_len:
+                if receiver_type == 1:
+                    assert not ('transparent' in result), "duplicate transparent receiver detected"
+                    assert len(acc) == 20
+                    result['transparent'] = bytes(acc)
+                    acc = []
+                    s = 0
+
+                elif receiver_type == 2:
+                    assert not ('sapling' in result), "duplicate sapling receiver detected"
+                    assert len(acc) == 43
+                    result['sapling'] = bytes(acc)
+                    acc = []
+                    s = 0
+
+                elif receiver_type == 3:
+                    assert not ('orchard' in result), "duplicate orchard receiver detected"
+                    assert len(acc) == 43
+                    result['orchard'] = bytes(acc)
+                    acc = []
+                    s = 0
+    return result
+
+
+def main():
+    args = render_args()
+
+    from random import Random
+    rng = Random(0xabad533d)
+    def randbytes(l):
+        ret = []
+        while len(ret) < l:
+            ret.append(rng.randrange(0, 256))
+        return bytes(ret)
+    rand = Rand(randbytes)
+
+    test_vectors = []
+    for _ in range(0, 10):
+        has_t_addr = rand.bool()
+        if has_t_addr:
+            t_addr = b"".join([rand.b(20)])
+        else:
+            t_addr = None
+
+        has_s_addr = rand.bool()
+        if has_s_addr:
+            sapling_sk = sapling_key_components.SpendingKey(rand.b(32))
+            sapling_default_d = sapling_sk.default_d()
+            sapling_default_pk_d = sapling_sk.default_pkd()
+            sapling_raw_addr = b"".join([sapling_default_d[:11], bytes(sapling_default_pk_d)[:32]])
+        else:
+            sapling_raw_addr = None
+
+        has_o_addr = (not has_s_addr) or rand.bool()
+        if has_o_addr:
+            orchard_sk = orchard_key_components.SpendingKey(rand.b(32))
+            orchard_fvk = orchard_key_components.FullViewingKey(orchard_sk)
+            orchard_default_d = orchard_fvk.default_d()
+            orchard_default_pk_d = orchard_fvk.default_pkd()
+            orchard_raw_addr = b"".join([orchard_default_d[:11], bytes(orchard_default_pk_d)[:32]])
+        else: 
+            orchard_raw_addr = None
+
+        receivers = {
+            'orchard': orchard_raw_addr, 
+            'sapling': sapling_raw_addr, 
+            'transparent': t_addr
+        }
+        ua = encode_unified(receivers)
```

Comment:
Using a dictionary for this makes the precise ordering of the UA internals non-deterministic (because Python uses a hashmap for dictionaries), which means non-deterministic test vectors. Use a list of tuples instead:
```suggestion
        receivers = [
            ('orchard', orchard_raw_addr),
            ('sapling', sapling_raw_addr),
            ('transparent', t_addr),
        ]
        ua = encode_unified(receivers)
```